prototype/mobile_endpoint/shardedmodels.py

```diff
@@ -0,0 +1,410 @@
+        # TODO: Select a shard by hashing the id
+
+    @classmethod
+    def get_shard_ids(cls, domain):
+        """
+        Return a list of shard ids where CaseData for the given domain should reside.
+        """
+        return [1, 2]
+        # TODO: Get this from a conf
+
+    @classmethod
+    def get_base_query(cls, session, shard_ids):
+        """
+        Return a query on this class filtered for the given shard ids.
+        """
+        return session.query(cls).filter(cls.shard_id.in_(shard_ids))
+
+
+# TODO: This chunk related to the sqlalchemy relationship between forms and cases
+# case_form_link = Table('case_form', Base.metadata,
+#     Column('case_id', UUID(), ForeignKey('case_data.id'), primary_key=True),
+#     Column('form_id', UUID(), ForeignKey('form_data.id'), primary_key=True)
+# )
+
+class CaseData(Base, ShardedModelMixin, ToFromGeneric):
+
+    __tablename__ = 'case_data'
+    id = Column(UUID(), primary_key=True)
+    shard_id = Column(SmallInteger, nullable=False)
+    domain = Column(Text(), nullable=False)
+    closed = Column(Boolean(), default=False, nullable=False)
+    owner_id = Column(UUID(), nullable=False)
+    server_modified_on = Column(DateTime(), nullable=False)
+    version = Column(Integer(), default=0)
+    case_json = Column(JSONB(), nullable=False)
+    attachments = Column(JSONB())
+
+    # TODO: Replace the functionality that this relationship provided
+    # forms = db.relationship("FormData", secondary=case_form_link, backref="cases")
+
+    def save(self):
+        assert self.id
+        assert self.domain
+        # TODO: Seems like there should be a more "automatic" way for setting this property
+        self.shard_id = self.get_shard_id(self.domain, self.id)
+        session = shard_manager.get_session(self.__class__, self.shard_id)
+        # TODO: Not sure what the implications of subtransactions are, but putting in here for now to make code run :p
+        with session.begin(subtransactions=True):
+            session.add(self)
+
+    @classmethod
+    def get_case(cls, domain, id):
+        shard = cls.get_shard_id(domain, id)
+        session = shard_manager.get_session(cls, shard)
+        query = cls.get_base_query(session, [shard])
+        return query.filter(cls.id == id).one()
+
+    @classmethod
+    def case_exists(cls, domain, id):
+        shard = cls.get_shard_id(domain, id)
+        session = shard_manager.get_session(cls, shard)
+        query = cls.get_base_query(session, [shard])
+        return query(exists().where(cls.id == id)).scalar()
+
+    @classmethod
+    def get_cases(cls, domain, case_ids, ordered=False):
+
+        shard_ids_by_session = defaultdict(list)
+        case_ids_by_shard = defaultdict(list)
+
+        for id in case_ids:
+            shard = cls.get_shard_id(domain, id)
+            session = shard_manager.get_session(cls, shard)
+            shard_ids_by_session[session].append(shard)
+            case_ids_by_shard[shard].append(id)
```

Comment:
This loop basically finds the shards that need to be queried and the sessions (specific database connections) that need to be queried.